cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -173,137 +151,133 @@ int ICACHE_RAM_ATTR stopWaveform(uint8_t pin) {
-    waveformState &= ~waveformToEnable;  // And clear the state of any just started
-    waveformToEnable = 0;
-    waveformToDisable = 0;
+    waveformsEnabled = (waveformsEnabled & ~waveformsToDisable) | waveformsToEnable; // Set the requested waveforms on/off
+    waveformsState &= ~waveformsToEnable;  // And clear the state of any just started
+    waveformsToEnable = 0;
+    waveformsToDisable = 0;
     // Find the first GPIO being generated by checking GCC's find-first-set (returns 1 + the bit of the first 1 in an int32_t)
-    startPin = __builtin_ffs(waveformEnabled) - 1;
+    startPin = __builtin_ffs(waveformsEnabled) - 1;
     // Find the last bit by subtracting off GCC's count-leading-zeros (no offset in this one)
-    endPin = 32 - __builtin_clz(waveformEnabled);
+    endPin = 32 - __builtin_clz(waveformsEnabled);
   }
 
   bool done = false;
-  if (waveformEnabled) {
+  if (waveformsEnabled) {
+    uint32_t now = ESP.getCycleCount();
     do {
-      nextEventCycles = microsecondsToClockCycles(MAXIRQUS);
+      nextEventCycle = now + microsecondsToClockCycles(MAXIRQUS);
       for (int i = startPin; i <= endPin; i++) {
-        uint32_t mask = 1<<i;
+        uint32_t mask = 1UL<<i;
 
         // If it's not on, ignore!
-        if (!(waveformEnabled & mask)) {
+        if (!(waveformsEnabled & mask))
           continue;
-        }
 
-        Waveform *wave = &waveform[i];
-        uint32_t now = GetCycleCountIRQ();
+        Waveform *wave = &waveforms[i];
 
-        // Disable any waveforms that are done
-        if (wave->expiryCycle) {
-          int32_t expiryToGo = wave->expiryCycle - now;
-          if (expiryToGo < 0) {
-              // Done, remove!
-              waveformEnabled &= ~mask;
-              if (i == 16) {
-                GP16O &= ~1;
-              } else {
-                ClearGPIO(mask);
-              }
-              continue;
-            }
+        // Check for toggles etc.
+        if (ExpiryState::UPDATE == wave->hasExpiry) {
+          wave->expiryCycle += wave->nextServiceCycle; // in ExpiryState::UPDATE, expiryCycle temporarily holds relative cycle count
+          if (waveformsState & mask)
+            wave->expiryCycle += wave->nextTimeLowCycles; // update expiry time to next full period
+          wave->hasExpiry = ExpiryState::ON;
         }
-
-        // Check for toggles
-        int32_t cyclesToGo = wave->nextServiceCycle - now;
-        if (cyclesToGo < 0) {
-          waveformState ^= mask;
-          if (waveformState & mask) {
+        const int32_t cyclesToGo = wave->nextServiceCycle - now;
+        int32_t expiryToGo = (ExpiryState:: ON == wave->hasExpiry) ? wave->expiryCycle - now : (cyclesToGo + 1);
+        const int32_t nextEventCycles = nextEventCycle - now;
+        if (cyclesToGo >= expiryToGo) {
+          // don't update waveform on or after expiring
+          expiryToGo = 0;
```

Comment:
I don't think I'm seeing the waveformEnabled mask changing when the expiry triggers.  Shouldn't the pin be removed from consideration once it's timed out?

Reply:
Further down below in the code, yes, it does.